### PR TITLE
Add support for Sparkfun Artemis Redboard

### PR DESCRIPTION
**Pull request by Topo-png**
I've attempted to implement the changes required to support the Sparkfun Artemis Reboard. I'm hung up trying to debug the following. 

```
In file included from C:\Users\1\Arduino\libraries\USB_Host_Shield_Library_2.0/Usb.h:43,
                 from C:\Users\1\Arduino\libraries\USB_Host_Shield_Library_2.0/XBOXONE.h:26,
                 from C:\Users\user1\AppData\Local\Temp\arduino_modified_sketch_523678\XBOXONE.ino:15:
C:\Users\1\Arduino\libraries\USB_Host_Shield_Library_2.0/usbhost.h:126:25: error: type/value mismatch at argument 1 in template parameter list for 'template<class SPI_CLK, class SPI_MOSI, class SPI_MISO, class SPI_SS> class SPi'
 typedef SPi< 5, 7, 6, 13> spi;
                         ^
C:\Users\1\Arduino\libraries\USB_Host_Shield_Library_2.0/usbhost.h:126:25: note:   expected a type, got '5'
C:\Users\1\Arduino\libraries\USB_Host_Shield_Library_2.0/usbhost.h:126:25: error: type/value mismatch at argument 2 in template parameter list for 'template<class SPI_CLK, class SPI_MOSI, class SPI_MISO, class SPI_SS> class SPi'
C:\Users\1\Arduino\libraries\USB_Host_Shield_Library_2.0/usbhost.h:126:25: note:   expected a type, got '7'
C:\Users\1\Arduino\libraries\USB_Host_Shield_Library_2.0/usbhost.h:126:25: error: type/value mismatch at argument 3 in template parameter list for 'template<class SPI_CLK, class SPI_MOSI, class SPI_MISO, class SPI_SS> class SPi'
C:\Users\1\Arduino\libraries\USB_Host_Shield_Library_2.0/usbhost.h:126:25: note:   expected a type, got '6'
C:\Users\1\Arduino\libraries\USB_Host_Shield_Library_2.0/usbhost.h:126:25: error: type/value mismatch at argument 4 in template parameter list for 'template<class SPI_CLK, class SPI_MOSI, class SPI_MISO, class SPI_SS> class SPi'
C:\Users\1\Arduino\libraries\USB_Host_Shield_Library_2.0/usbhost.h:126:25: note:   expected a type, got '13'
C:\Users\1\Arduino\libraries\USB_Host_Shield_Library_2.0/usbhost.h: In member function 'int8_t MAX3421e<SPI_SS, INTR>::Init()':
C:\Users\1\Arduino\libraries\USB_Host_Shield_Library_2.0/usbhost.h:434:18: error: qualified-id in declaration before '(' token
         spi::init();
                  ^
C:\Users\1\Arduino\libraries\USB_Host_Shield_Library_2.0/usbhost.h: In member function 'int8_t MAX3421e<SPI_SS, INTR>::Init(int)':

C:\Users\1\Arduino\libraries\USB_Host_Shield_Library_2.0/usbhost.h:471:18: error: qualified-id in declaration before '(' token
         spi::init();
                  ^
In file included from C:\Users\1\Arduino\libraries\USB_Host_Shield_Library_2.0/Usb.h:44,
                 from C:\Users\1\Arduino\libraries\USB_Host_Shield_Library_2.0/XBOXONE.h:26,
                 from C:\Users\user1\AppData\Local\Temp\arduino_modified_sketch_523678\XBOXONE.ino:15:
C:\Users\1\Arduino\libraries\USB_Host_Shield_Library_2.0/UsbCore.h: At global scope:
C:\Users\1\Arduino\libraries\USB_Host_Shield_Library_2.0/UsbCore.h:212:29: error: expected class-name before '{' token
 class USB : public MAX3421E {
                             ^
Using library SPI at version 1.0 in folder: C:\Users\user1\AppData\Local\Arduino15\packages\SparkFun\hardware\apollo3\1.2.1\libraries\SPI 

Using library USB_Host_Shield_Library_2.0 at version 1.6.0 in folder: C:\Users\1\Arduino\libraries\USB_Host_Shield_Library_2.0 

exit status 1

Error compiling for board SparkFun RedBoard Artemis ATP.
```

### CODE REVIEW

File: avrpins.h [hunk 1]
```diff
@@ -1356,6 +1356,60 @@ MAKE_PIN(P33, (9));
...
+#if defined(AM_PACKAGE_BGA)
+	#define P38 38
+	#define P39 39
+	#define P40 40
+	#define P41 41
+	#define P42 42
+	#define P43 43
+	#define P44 44
+	#define P45 45
+#endif
```
Comment (Lauszus) [inline avrpins.h:1411]: You need create a `MAKE_PIN` macro and use that to define the pins. See the `MIPS` port for some inspiration: <https://github.com/felis/USB_Host_Shield_2.0/blob/master/avrpins.h#L1479-L1521>

File: avrpins.h [hunk 2]
```diff
@@ -1356,6 +1356,60 @@ MAKE_PIN(P33, (9));
 
 #undef MAKE_PIN
 
+#elif defined(AM_PART_APOLLO3)
+// Sparkfun Artemis Redboard & ATP variant
+#include "Arduino.h" \
```
Comment (Lauszus) [inline avrpins.h:1361]: Remove this include.

File: UsbCore.h [hunk 3]
```diff
@@ -52,6 +52,8 @@ typedef MAX3421e<P15, P5> MAX3421E; // ESP8266 boards
 typedef MAX3421e<P5, P17> MAX3421E; // ESP32 boards
 #elif (defined(__AVR_ATmega644P__) || defined(__AVR_ATmega1284P__))
 typedef MAX3421e<Pb4, Pb3> MAX3421E; // Sanguino
+#elif defined(AM_PART_APOLLO3)
+typedef MAX3421e<P10, P9> MAX3421E; // Sparkfun Redboard Artemis All the pins
```
Comment (Lauszus) [inline UsbCore.h:56]: No need to add this, as pin 10 and 9 are already the default i.e. remove these two lines.

File: settings.h [hunk 4]
```diff
@@ -175,6 +175,10 @@ e-mail   :  support@circuitsathome.com
 extern SPI_HandleTypeDef SPI_Handle; // Needed to be declared in your main.cpp
 #endif
 
+#ifdef AM_PART_APOLLO3
+#include <SPI.h> 
+#endif
+
```
Comment (Lauszus) [inline settings.h:181]: No need to add this include.

File: usbhost.h [hunk 5]
```diff
@@ -122,6 +122,8 @@ typedef SPi< P14, P13, P12, P15 > spi;
 typedef SPi< P18, P23, P19, P5 > spi;
 #elif defined(ARDUINO_NRF52840_FEATHER)
 typedef SPi< P26, P25, P24, P5 > spi;
+#elif defined(AM_PART_APOLLO3)
+typedef SPi< P5, P7, P6, P13> spi;
```
Comment (Lauszus) [inline usbhost.h:126]: This should be: 

```C++
typedef SPi< P13, P11, P12, P10 > spi;
```

File: avrpins.h (hunk 1 above)
Comment (Topo-png) [reply avrpins.h:1411]: Thanks, I thought I would be able to define the pins similar to other Arudino variants.

File: UsbCore.h (hunk 3 above)
Comment (Topo-png) [reply UsbCore.h:56]: Thanks, I had put this in there in case I needed to change it later on. Removed

File: settings.h (hunk 4 above)
Comment (Topo-png) [reply settings.h:181]: Removed

File: usbhost.h (hunk 5 above)
Comment (Topo-png) [reply usbhost.h:126]: Thanks for the insight.

File: avrpins.h (hunk 1 above)
Comment (Topo-png) [reply avrpins.h:1428]: Thanks, I copied the MIPS port and it seems to build properly.

File: avrpins.h (hunk 2 above)
Comment (Topo-png) [reply avrpins.h]: Removed

**Comment by Topo-png** [comment]
It looks to build successfully on my machine @Lauszus. Thanks for the guidance. This PR was my first and also my first real attempt at using Git.

File: avrpins.h [hunk 6]
```diff
@@ -1356,6 +1356,77 @@ MAKE_PIN(P33, (9));
...
+MAKE_PIN(P39, 39);
+#ifdef AM_PACKAGE_BGA // All The Pins (ATP) variant
+MAKE_PIN(P40, 40);
+MAKE_PIN(P41, 41);
+MAKE_PIN(P42, 42);
+MAKE_PIN(P43, 43);
+MAKE_PIN(P44, 44);
+MAKE_PIN(P45, 45);
+#endif
+
```
Comment (Lauszus) [inline avrpins.h:1429]: Please add:

```C++
#undef MAKE_PIN
```

File: usbhost.h [hunk 7]
```diff
@@ -429,7 +436,7 @@ int8_t MAX3421e< SPI_SS, INTR >::Init() {
         /* pin and peripheral setup */
         SPI_SS::SetDirWrite();
         SPI_SS::Set();
-        spi::init();
+		spi::init();
```
Comment (Lauszus) [inline usbhost.h:439]: Please undo this change.

Comment (Topo-png) [reply usbhost.h]: Weird, this didn't show properly in Notepad++. Updated

File: usbhost.h [hunk 8]
```diff
@@ -57,6 +57,10 @@ template< typename SPI_CLK, typename SPI_MOSI, typename SPI_MISO, typename SPI_S
         static void init() {
                 // Should be initialized by the user manually for now
         }
+#elif defined(AM_PART_APOLLO3)
+		static void init() {
+				SPI::begin();
+		}
```
Comment (Lauszus) [inline usbhost.h:63]: Please remove this.

Comment (Lauszus) [reply usbhost.h:63]: Instead add the following to the bottom of [settings.h](https://github.com/felis/USB_Host_Shield_2.0/blob/b56e8b76a1bf1b93be8c82232e34c9c17d799bbd/settings.h#L217):

```C++
// See: https://github.com/sparkfun/Arduino_Apollo3/pull/421
#ifdef AM_PART_APOLLO3
#ifndef SPI_HAS_TRANSACTION
#define SPI_HAS_TRANSACTION 1
#endif
#endif
```

Comment (Topo-png) [reply usbhost.h]: removed, and updated

File: settings.h [hunk 9]
```diff
@@ -216,3 +216,10 @@ extern SPI_HandleTypeDef SPI_Handle; // Needed to be declared in your main.cpp
 #endif
 
 #endif /* SETTINGS_H */
+
+// See: https://github.com/sparkfun/Arduino_Apollo3/pull/421
+#ifdef AM_PART_APOLLO3
+#ifndef SPI_HAS_TRANSACTION
+#define SPI_HAS_TRANSACTION 1
+#endif
+#endif
```
Comment (Lauszus) [inline settings.h:225]: You need to add this before `#endif /* SETTINGS_H */` :)

Comment (Topo-png) [reply settings.h:223]: whoops... corrected

**Comment by Lauszus** [comment]
All looks good now. Have you confirmed that the library now works with the board?

**Comment by Topo-png** [comment]
I'm waiting for a new USB host shield to arrive. Examples build properly. I will hopefully be able to report functionality later this week.

**Comment by Lauszus** [comment]
@Topo-png thanks. I'll merge and squash the commits once you have verified that it is actually working with the board :)

**Comment by Topo-png** [comment]
I loaded XBOXUSB example to the board for testing. 
Serial Log reports back
`OSC did not start`

**Comment by xxxajk** [comment]
Use shorter wires for SPI. :-)

**Comment by Topo-png** [comment]
@xxxajk - this is a shield directly connected to header pins, no long wires other than the PCB traces. 

I modified the pins to match what I believe is the correct pinout. 

Board_qc now outputs the following
`Circuits At Home 2011
USB Host Shield Quality Control Routine
Reading REVISION register... Die revision 03
SPI long test. Transfers 1MB of data. Each dot is 64K
Test failed.  Value written: 01 read: 00
Unrecoverable error - test halted!!
0x55 pattern is transmitted via SPI
Press RESET to restart test`

**Comment by xxxajk** [comment]
Slow down the SPI, going at maximum rated rate with some SPI causes issues. This was noted before on Atmel's SAMD ARM, where it's SPI didn't clock the bit in at the correct point, and would miss or hit it in transition.
What speed are you currently using?

**Comment by Topo-png** [comment]
I haven't changed any SPI speeds. How do I slow this down?

**Comment by xxxajk** [comment]
Unfortunately someone (not me) didn't make it a define so... Here's the list.

https://github.com/felis/USB_Host_Shield_2.0/blob/master/usbhost.h#L182
https://github.com/felis/USB_Host_Shield_2.0/blob/master/usbhost.h#L225
https://github.com/felis/USB_Host_Shield_2.0/blob/master/usbhost.h#L283
https://github.com/felis/USB_Host_Shield_2.0/blob/master/usbhost.h#L322

**Comment by Topo-png** [comment]
Thanks, I reduced those down to 260000 and I still didn't have any luck. The errors take noticeably longer to report so I believe this is working properly.

**Comment by xxxajk** [comment]
Try the following:
12000000
10000000
8000000

If you can't get it working at less than 8MHz, then something else is wrong.
Running the shield on 3.3v? Because 5V will fry the esp...

**Comment by xxxajk** [comment]
One other note, is that the official shields, use the ICSP header for SPI. Just sayin'...
